### ISSUE TITLE
If a cancellation function for a Deferred throws an exception, behavior is undefined

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6676]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#6676|
|Type|defect|
|Created|2013-08-12 13:36:49Z|

In general the contract for Deferred cancellation is that it should always succeed, and calling `cancel()` multiple times should have no effect. However, if the cancellation function throws an exception it is currently not caught, and cancellation does not occur.

One of the following options needs to be chosen, I think:

 1. Catch the exception and log it, and fallback to just firing `Deferred` with `CancelledError`.
 2. Document this particular case as the one situation where `cancel()` may raise an exception.

<details><summary>Searchable metadata</summary>

```
trac-id__6676 6676
type__defect defect
reporter__itamar itamar
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__new new
resolution__None None
component__core core
keywords__ 
time__1376314609000000 1376314609000000
changetime__1379279594000000 1379279594000000
version__None None
owner__kaizhang kaizhang
cc__glyph
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

Calling `cancel()` multiple times cancels the waited-upon Deferred multiple times; the contract is not that it will "always succeed" (what does "succeed" even mean in this context?) "Contract" sounds like a formalism, so it would be good to be formal in your definitions here.

(I just wrote a monster mailing-list post about this, so more commentary is probably superfluous.)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst commented|
|-|-|

After some discussion on mailing list and IRC it was concluded that best solution is to log the exception, and then fire the Deferred with `CancelledError`. The exception indicates a bug in the cancellation function, we really shouldn't be preventing the Deferred from firing, given that one of the goals of cancellation is to allow someone waiting on a Defererd to proceed. A bug shouldn't prevent that.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

Replying to [itamar](#note_2):
> After some discussion on mailing list and IRC it was concluded that best solution is to log the exception, and then fire the Deferred with `CancelledError`. The exception indicates a bug in the cancellation function, we really shouldn't be preventing the Deferred from firing, given that one of the goals of cancellation is to allow someone waiting on a Defererd to proceed. A bug shouldn't prevent that.

I agree completely, and this is pretty much exactly what I said in my mailing list post.  Which is [here on GMane](http://news.gmane.org/gmane.comp.python.twisted), for posterity.
